File: web/src/services/StatisticsService.ts
import { get } from "@/utils/fetch";
import type { StatRange } from "@/pages/dashboard/types";

const baseUrl = "/api/v1/statistics";

type Opts = { signal?: AbortSignal };

export type RankingType =
  | "host"
  | "path"
  | "referer_host"
  | "referer_url"
  | "os"
  | "browser"
  | "status";

const qs = (params: Record<string, string | number | boolean | undefined>) =>
  Object.entries(params)
    .filter(([, v]) => v !== undefined && v !== "")
    .map(([k, v]) => `${k}=${encodeURIComponent(String(v))}`)
    .join("&");

export const getStatisticsOverview = (range: StatRange, host?: string, opts?: Opts) =>
  get(`${baseUrl}/overview?${qs({ range, host })}`, opts);

export const getStatisticsTimeseries = (range: StatRange, host?: string, opts?: Opts) =>
  get(`${baseUrl}/timeseries?${qs({ range, host })}`, opts);

export const getStatisticsGeo = (
  range: StatRange,
  scope: "world" | "china",
  mode: "all" | "blocked",
  host?: string,
  top = 20,
  opts?: Opts
) => get(`${baseUrl}/geo?${qs({ range, scope, mode, host, top })}`, opts);

export const getStatisticsRankings = (
  range: StatRange,
  type: RankingType,
  host?: string,
  top = 10,
  mode: "all" | "blocked" = "all",
  opts?: Opts
) => get(`${baseUrl}/rankings?${qs({ range, type, host, top, mode })}`, opts);

export const getStatisticsRankingsPaged = (
  range: StatRange,
  type: RankingType,
  page: number,
  pageSize: number,
  host?: string,
  opts?: Opts
) => get(`${baseUrl}/rankings?${qs({ range, type, page, pageSize, host })}`, opts);

export const getStatisticsRequests = (
  range: StatRange,
  params: { host?: string; ip?: string; status?: number; blocked?: boolean; page?: number; pageSize?: number },
  opts?: Opts
) => get(`${baseUrl}/requests?${qs({ range, ...params })}`, opts);

File: src/FastGateway/Services/Statistics/StatisticsService.cs
using FastGateway.Infrastructure;

namespace FastGateway.Services.Statistics;

public static class StatisticsService
{
    public static IEndpointRouteBuilder MapStatistics(this IEndpointRouteBuilder app)
    {
        // 查询侧懒初始化（后台服务未跑起来之前也能安全响应）
        StatisticsDb.Initialize();

        var statistics = app.MapGroup("/api/v1/statistics")
            .WithTags("统计分析")
            .WithDescription("流量统计分析")
            .RequireAuthorization()
            .AddEndpointFilter<ResultFilter>()
            .WithDisplayName("统计分析");

        statistics.MapGet("overview", (string? range, string? host) =>
                StatisticsQueryService.GetOverview(range, host))
            .WithDescription("统计概览")
            .WithDisplayName("统计概览");

        statistics.MapGet("timeseries", (string? range, string? host) =>
                StatisticsQueryService.GetTimeSeries(range, host))
            .WithDescription("时间序列")
            .WithDisplayName("时间序列");

        statistics.MapGet("geo", (string? range, string? host, string? scope, string? mode, int? top) =>
                StatisticsQueryService.GetGeo(range, host, scope, mode, top ?? 20))
            .WithDescription("地理分布")
            .WithDisplayName("地理分布");

        statistics.MapGet("rankings",
                (string? range, string? host, string? type, string? mode, int? top, int? page, int? pageSize) =>
                StatisticsQueryService.GetRankings(range, host, type, mode, top ?? 10, page, pageSize))
            .WithDescription("维度排行")
            .WithDisplayName("维度排行");

        statistics.MapGet("requests",
                (string? range, string? host, string? ip, int? status, bool? blocked, int? page, int? pageSize) =>
                StatisticsQueryService.GetRequests(range, host, ip, status, blocked, page ?? 1, pageSize ?? 20))
            .WithDescription("请求明细")
            .WithDisplayName("请求明细");

        return app;
    }
}
